plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/ExpEmbedding.kt

```diff
@@ -129,71 +144,76 @@ sealed interface BinaryBooleanExpression : ExpEmbedding {
 data class And(
     override val left: ExpEmbedding,
     override val right: ExpEmbedding,
+    override val source: KtSourceElement? = null,
 ) : BinaryBooleanExpression {
-    override fun toViper() = Exp.And(left.toViper(), right.toViper())
+    override fun toViper() = Exp.And(left.toViper(), right.toViper(), source.asSilverPosition)
 }
 
 data class Or(
     override val left: ExpEmbedding,
     override val right: ExpEmbedding,
+    override val source: KtSourceElement? = null,
 ) : BinaryBooleanExpression {
-    override fun toViper() = Exp.Or(left.toViper(), right.toViper())
+    override fun toViper() = Exp.Or(left.toViper(), right.toViper(), source.asSilverPosition)
 }
 
 data class Implies(
     override val left: ExpEmbedding,
     override val right: ExpEmbedding,
+    override val source: KtSourceElement? = null,
 ) : BinaryBooleanExpression {
-    override fun toViper() = Exp.Implies(left.toViper(), right.toViper())
+    override fun toViper() = Exp.Implies(left.toViper(), right.toViper(), source.asSilverPosition)
 }
 
 data class Not(
     val exp: ExpEmbedding,
+    override val source: KtSourceElement? = null,
 ) : ExpEmbedding {
     override val type = BooleanTypeEmbedding
 
-    override fun toViper() = Exp.Not(exp.toViper())
+    override fun toViper() = Exp.Not(exp.toViper(), source.asSilverPosition)
 }
 
 
 data object UnitLit : ExpEmbedding {
     override val type = UnitTypeEmbedding
+    override val source: KtSourceElement? = null
 
     override fun toViper() = UnitDomain.element
 }
 
-data class IntLit(val value: Int) : ExpEmbedding {
+data class IntLit(val value: Int, override val source: KtSourceElement? = null) : ExpEmbedding {
     override val type = IntTypeEmbedding
 
     override fun toViper() = Exp.IntLit(value)
 }
 
-data class BooleanLit(val value: Boolean) : ExpEmbedding {
+data class BooleanLit(val value: Boolean, override val source: KtSourceElement? = null) : ExpEmbedding {
     override val type = BooleanTypeEmbedding
 
     override fun toViper() = Exp.BoolLit(value)
 }
 
-data class NullLit(val elemType: TypeEmbedding) : ExpEmbedding {
+data class NullLit(val elemType: TypeEmbedding, override val source: KtSourceElement? = null) : ExpEmbedding {
     override val type = NullableTypeEmbedding(elemType)
-
-    override fun toViper() = NullableDomain.nullVal(elemType.viperType)
+    override fun toViper() = NullableDomain.nullVal(elemType.viperType, source)
 }
 
-data class Is(val exp: ExpEmbedding, val comparisonType: TypeEmbedding) : ExpEmbedding {
+data class Is(val exp: ExpEmbedding, val comparisonType: TypeEmbedding, override val source: KtSourceElement? = null) : ExpEmbedding {
```

Comment:
Is the `source` intentionally not used here?

Reply:
Yes, the source is propagated by the given `exp` field